### ISSUE TITLE
If a role is skipped due to failed conditional, the role's dependencies are skipped in subsequent calls

### ISSUE THREAD

**Issue by nizig** [ansible/ansible#14438]
### Version

ansible 2.0.0.2
### Test

```
------ inventory ------

[local]
127.0.0.1


------ playbook.yml ------

---

- hosts: local
  gather_facts: no
  roles:
    - role: alpha
      when: False
    - alpha

------ roles/alpha/meta/main.yml ------

---

allow_duplicates: yes
dependencies:
  - beta

------ roles/alpha/tasks/main.yml ------

---

- debug: msg=alpha

------ roles/beta/tasks/main.yml ------

---

- debug: msg=beta
```
### Invocation

```
$ ansible-playbook -i inventory playbook.yml
```
### Expected result

```
PLAY ***************************************************************************

TASK [beta : debug] ************************************************************
skipping: [127.0.0.1]

TASK [alpha : debug] ***********************************************************
skipping: [127.0.0.1]

TASK [beta : debug] ************************************************************
ok: [127.0.0.1] => {
    "msg": "beta"
}

TASK [alpha : debug] ***********************************************************
ok: [127.0.0.1] => {
    "msg": "alpha"
}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0   
```
### Actual result

```
PLAY ***************************************************************************

TASK [beta : debug] ************************************************************
skipping: [127.0.0.1]

TASK [alpha : debug] ***********************************************************
skipping: [127.0.0.1]

TASK [beta : debug] ************************************************************
skipping: [127.0.0.1]

TASK [alpha : debug] ***********************************************************
ok: [127.0.0.1] => {
    "msg": "alpha"
}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=1    changed=0    unreachable=0    failed=0   
```

**Comment by khadersyed** [comment]
We're facing the same problem too. Can we expect a fix in the near future?

**Comment by rodnaph** [comment]
This same bug was found and reported as fixed in <2.0, maybe a regression: https://github.com/ansible/ansible/issues/7353

**Comment by marcpo** [comment]
Also having the same issue. Seems this code is causing this behavior. In `lib/ansible/playbook/role/__init__.py`:

```

            if parent_role:
                if parent_role.when:
                    new_when = parent_role.when[:]
                    new_when.extend(r.when or [])
                    r.when = new_when
```

I also have different repro step related to conditionals in roles

```
------ roles/r1/meta/main.yml ------
- debug: msg="running r1"

------ roles/r2/meta/main.yml ------
dependencies:
  - { role: r1}

------ roles/r3/meta/main.yml ------
dependencies:
  - { role: r2, when: false }

------ roles/r4/meta/main.yml ------
dependencies:
  - { role: r1 }

---------- playbook1.yml ---------------
- hosts: localhost
  roles:
    - { role: r3 }
    - { role: r4 }

```

---------- Run it as:---------------------
ansible-playbook -i localhost, playbook1.yml -v

The results: r1 is always skipped

> No config file found; using defaults
> 
> PLAY [localhost] ***************************************************************
> 
> TASK [setup] *******************************************************************
> ok: [localhost]
> 
> TASK [r1 : debug] **************************************************************
> **skipping: [localhost] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}**
> 
> TASK [r1 : debug] **************************************************************
> **skipping: [localhost] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}**
> 
> PLAY RECAP *********************************************************************
> localhost                  : ok=1    changed=0    unreachable=0    failed=0

Will get the proper result if the roles are inverted in playbook1.yml

```
---------- playbook2.yml ---------------
- hosts: localhost
  roles:
    - { role: r4 }
    - { role: r3 }

```

The results: r1 is called

> No config file found; using defaults
> 
> PLAY [localhost] ***************************************************************
> 
> TASK [setup] *******************************************************************
> ok: [localhost]
> 
> **TASK [r1 : debug] **************************************************************
> ok: [localhost] => {
>     "msg": "running r1"
> }**
> 
> PLAY RECAP *********************************************************************
> localhost                  : ok=2    changed=0    unreachable=0    failed=0

**Comment by nhooey** [comment]
I can confirm this behaviour in Ansible 2.0.0.2 and Ansible 2.0.2.0.
@jimi-c, @mpdehaan

**Comment by sidcarter** [comment]
I can confirm this behaviour with Ansible 2.2.0 too.

**Comment by camhart** [comment]
Can anyone confirm if this is fixed in ansible v2.1?

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release (2.1.1).

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
